src/Spec.ts

```diff
@@ -938,15 +981,72 @@ export default class Spec {
       // @ts-ignore
       let grammars = [...sdo.children].filter(e => e.tagName === 'EMU-GRAMMAR');
       for (let grammar of grammars) {
-        for (let rhs of grammar.querySelectorAll('emu-rhs')) {
-          let alternativeId = rhs.getAttribute('a');
-          if (!{}.hasOwnProperty.call(sdoMap, alternativeId)) {
-            sdoMap[alternativeId] = [];
+        for (let production of (grammar.querySelectorAll('emu-production') as any as Iterable<Element>)) {
+          if (!production.hasAttribute('name')) {
+            // I don't think this is possible, but we can at least be graceful about it
+            this.warn({
+              type: 'node',
+              node: grammar, 
+              ruleId: 'grammar-shape',
+              message: 'expected emu-production node to have name',
+            });
+            continue;
+          }
+          let name = production.getAttribute('name')!;
+          if (!productions.has(name)) {
+            this.warn({
+              type: 'node',
+              node: grammar,
+              ruleId: 'grammar-shape',
+              message: `could not find definition corresponding to production ${name}`,
+            });
+            continue;
+          }
+          let mainRhses = productions.get(name)!;
+          for (let rhs of (production.querySelectorAll('emu-rhs') as any as Iterable<Element>)) {
+            let matches = mainRhses.filter(r => rhsMatches(rhs, r));
+            if (matches.length === 0) {
+              this.warn({
+                type: 'node',
+                node: grammar,
+                ruleId: 'grammar-shape',
+                message: `could not find definition for rhs ${rhs.textContent}`,
+              });
+              continue;
+            }
+            if (matches.length > 1) {
+              this.warn({
+                type: 'node',
+                node: grammar,
+                ruleId: 'grammar-shape',
+                message: `found multiple definitions for rhs ${rhs.textContent}`,
+              });
+              continue;
+            }
+            let match = matches[0];
+            if (match.id == '') {
+              match.id = 'prod-' + sha(`${name} : ${match.textContent}`);
+            }
+            let mainId = match.id;
+            if (rhs.id == '') {
+              rhs.id = 'prod-' + sha(`[${sdoName}] ${name} ${rhs.textContent}`);
```

Comment:
You know I'd prefer this to be an hmac. This should be fine, though, if we don't have an hmac function handy.